### PR TITLE
Add BLENDED model

### PR DESCRIPTION
**Pull request by julemai**
The BLENDED model allows for simultaneous calibration of model parameters and model structure (see [Mai et al. (2020)](https://doi.org/10.5194/hess-24-5835-2020)).

- Both tests (`TestBLENDED` and `TestBLENDED_OST`) pass.
- Test `test_blended.py` should be merged into `test_emulators.py`. Just did not work on my system due to missing GDAL and @Zeitsperre made test to skip in that case.
- Created branch off `raven_gis_ports`.

Things to discuss/ implement:
- [x] parameter avg basin slope [deg] `SLOPE` (RVH) should be templated (similar to, for example, `elevation`) 
- [x] parameter forest cover [frac] `FOREST_COV` (RVP) should be templated (similar to, for example, `elevation`)  
- [x] RVT file should contain two parameters `:RainCorrection` and `:SnowCorrection`. Can be default at 1.0 but might be given as parameter (here `par_x33` and `par_x34`, respectively)

### CODE REVIEW

**Comment by Zeitsperre** [comment]
@julemai Just to be sure that we don't pollute the commit history, be sure to perform a `Squash and Merge`, removing the history before commit `8f31afd`. Thanks!

File: ravenpy/models/emulators.py [hunk 1]
```diff
@@ -437,6 +439,192 @@ def __init__(self, *args, **kwds):
     def derived_parameters(self):
         self._monthly_average()
 
+class BLENDED(GR4JCN):
```
[the remark this answers is not in the document]
Comment (huard) [reply ravenpy/models/emulators.py]: Done in my commit.

File: ravenpy/models/emulators.py [hunk 2]
```diff
@@ -437,6 +439,192 @@ def __init__(self, *args, **kwds):
...
+        )
+
+    def derived_parameters(self):
+        self.rvd["TOPSOIL_hlf"] = self.rvp.params.par_x29 * 0.5 * 1000.0
+        self.rvd["PHREATIC_hlf"] = self.rvp.params.par_x30 * 0.5 * 1000.0
+        self.rvd["TOPSOIL_mm"] = self.rvp.params.par_x29 * 1000.0
+        self.rvd["PHREATIC_mm"] = self.rvp.params.par_x30 * 1000.0
+        self.rvd[
+            "SUM_X09_X10"
+        ] = self.rvp.params.par_x10  # + self.rvp.params.par_x09
```
[the remark this answers is not in the document]
Comment (julemai) [reply ravenpy/models/emulators.py:486]: I tried to be consistent with models before. HMETS had that comments too. It is basically that x10 is actually (x10+x09) and I think it is not so helpful to obtain that from `SUM_X09_X10`.

[the remark this answers is not in the document]
Comment (huard) [reply ravenpy/models/emulators.py:486]: Got it. 

In terms of calibration, I would assume that this kind of co-linearity can get problematic, no? Is it not better to define orthogonal parameters, then compute derived quantities from those ?

[the remark this answers is not in the document]
Comment (julemai) [reply ravenpy/models/emulators.py:486]: This is how we had that implemented initially. :D Then we switched to this setup. It is totally fine for me as it is. 

Ostrich is receiving the proper (non-derived) parameters and Raven is using the proper (derived) parameters.

[the remark this answers is not in the document]
Comment (huard) [reply ravenpy/models/emulators.py:486]: Ok, thanks !

**Comment by huard** [comment]
Hi @julemai 

Please take a look at the modifications I made to template the remaining parameters. In the last commit, the rain and snow correction values are taken from the parameter list, but this changes the model output and breaks the asserts. 

Otherwise, I think this is good to go. I think it will be useful to include for each emulator some documentation and references. Will open a PR for this (maintenance).

**Comment by julemai** [comment]
Ignore what I said... It's `para_x33` and `para_x34`. Test 1 is passing now and that means that I am VERY sure that the second one should also be correct. Just double-checking. One second....

**Comment by julemai** [comment]
The Ostrich test fails and this is because `:RainCorrection` and `:SnowCorrection` get filled in all the time with 1 (default). But they should be filled with `par_x33` and `par_x34` such that Ostrich knows where to fill in values during calibration. I just don't know how and where to adjust this. @huard could you have a look?

File: ravenpy/models/ostrich-blended/raven-blended.rvt.tpl [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#
+# Generic input file for global models with area averaged series
+#---------------------------------------------------------------
+
+:Gauge meteorological forcings
+   :Latitude    {latitude}
+   :Longitude   {longitude}
+   :Elevation   {elevation}
+   {raincorrection_cmd}
+   {snowcorrection_cmd}
```
Comment (huard) [inline ravenpy/models/ostrich-blended/raven-blended.rvt.tpl:14]: Could you just replace this with 
```
   :RainCorrection par_x33
   :SnowCorrection par_x34
``` ?
I've tried and test still fail on my end, but maybe I'm missing something else.

Comment (julemai) [reply ravenpy/models/ostrich-blended/raven-blended.rvt.tpl:14]: Yeah, I think it is expected that the test still fails. When I replace it with the above it at least fills in the proper values. Latitude, longitude and elevation are still `None`.

Comment (huard) [reply ravenpy/models/ostrich-blended/raven-blended.rvt.tpl:14]: I think I know. When calling the model, I've replaced latitude, longitude, elevation arguments by the HRU object, which includes these values and is used to fill the RVH. 
But these are not the same as the latitude, longitude, elevation of the `Gauge` object, which now never are instantiated. 

Up to now, we assumed these two sets of latitude, longitude, elevation were the same thing. This is not a correct assumption, right ?

What is a good solution to specify Gauge latitude, longitude, elevation ? 
- Extraction from the netCDF file (elevation will in most cases not be found)
- User-provided `Gauge` parameters 
- Taking the values from the HRU (status quo)

Comment (julemai) [reply ravenpy/models/ostrich-blended/raven-blended.rvt.tpl:14]: I think in our case it is not really a "gauge" but the basin averaged rain, snow, etc. Then I would provide there also the basin elevation, latitude, and longitude. But it probably won't matter too much.

**Comment by julemai** [comment]
Tests are both passing on my end. I confirmed that it is the same as `8f31afd` when `x33` and `x34` are fixed during calibration.

**Comment by huard** [comment]
Great !

Yes. Don't do anything fancy for now, maybe just one line about what is special about the model and the reference to your paper. 

Also, is "BLENDED" the official name of your model ? If there is no official name, I would suggest finding a more memorable name.

**Comment by huard** [comment]
@julemai it looks like in the Gauge command, I can leave the lat, lon, elev to None and all tests pass. But if remove the Commands altogether, then the model complains and fail. 

Given it does not seem to make any difference, I would revisit this issue later and not hold up this PR.

**Comment by julemai** [comment]
Yes, let's leave them to `None` then. The commands can not be removed; they should be set to the elevation etc you use at other parts too. But let's go ahead.

**Comment by huard** [comment]
I suggest we wait for the 0.3 release before merging into master. 
Note that we'll need to update HISTORY to reflect this new, very cool addition.

**Comment by cjauvin** [comment]
Please use: https://github.com/CSHS-CWRA/RavenPy/pull/76

**Pull request closed without merging**